### DISCUSSION TITLE
Field description as docstring

### DISCUSSION THREAD

**Discussion by tomhampshire** [pydantic/pydantic#4281] [Feature Request]
The [schema documentation](https://pydantic-docs.helpmanual.io/usage/schema/) states that:
"The description for models is taken from either the docstring of the class or the argument description to the Field class."
which results in models similar to:

_example 1_
```python
class Klass(BaseModel):
    """Klass docstring"""

    foo: str = Field(..., description="help text")

Klass(foo="test")
```

Now, when you hover over `Klass` in `Klass(foo="test")`, the docstring for `Klass` will be shown (using pyright/pylance). However, when hovering over `foo` in `Klass(foo="test")`, no docstring is shown. This can be remedied by the following:

_example 2_
```python
class Klass(BaseModel):
    """Klass docstring"""

    foo: str = Field(..., description="help text")
    """help text"""

Klass(foo="test")
```

However, you now have redundant code. Removing the `description` from the `Field` means that when you generate a schema, the description is missing:

_example 3_
```python
class Klass(BaseModel):
    """Klass docstring"""

    foo: str = Field(...)
    """help text"""

Klass(foo="test")
```

My feature proposal is to be able to use either the syntax from example 1 or 3 and retain both hover documentation for the Field *and* schema generation with the Field description

**Comment by shc261392** [comment]
I would second on this one. In most use cases, the docstring and `Field` description could use the same text.

Making the docstring a default value for `Field` description would make adding description to schema and docstring at the same time easier.

---

edit:

After taking a look at the `Field` and `FieldInfo` implementation, I couldn't figure out a way to implement the feature describe above without modifying too much code. Currently the `Field` function cannot access the original attribute nor its `__doc__` attribute and it is not possible to pass the field attribute to it in the same statement.

Please correct me if I'm wrong, but it seems the attributes' `__doc__` field is also not stored anywhere in the `BaseModel` class. So it seems creating a custom class inheriting the `BaseModel` class and load the docstring into description is not as easy as it seems.

**Comment by Viicos** [comment]
Unfortunately there's currently no way of using the docstring from the class attributes (see this [stackoverflow](https://stackoverflow.com/questions/55672640/how-can-i-get-doc-string-of-the-class-attribute-in-python) thread). Afaik `__doc__` is only available for functions/methods and classes (and maybe some other edge cases).

**Reply by adriangb** [reply]
Yes, this is the case and unless Python changes this there is nothing Pydantic can do. @tiangolo does have something in the works to solve this another way though.

**Reply by Viicos** [reply]
Yes I just learned about [`inspect.getsourcelines`](https://docs.python.org/3/library/inspect.html#inspect.getsourcelines), used by [`class_doc`](https://github.com/danields761/class-doc). More on this here: https://github.com/pydantic/pydantic-settings/issues/2. Indeed having this built in `pydantic` would be awesome, altough using `inspect` seems a bit hacky

**Reply by apirogov** [reply]
While this is still `inspect`-based under the hood, [this](https://github.com/lebrice/SimpleParsing/blob/master/simple_parsing/docstring.py) is useful and I used `get_attribute_docstring` to "hack" this feature into my pydantic (V1) models.

Would also be really happy if this would be supported, somehow. Maybe on opt-in basis (if it impacts performance too negatively).

**Reply by Viicos** [reply]
@apirogov See https://github.com/pydantic/pydantic/pull/6563 for additional discussion. I'll get back on it this week probably

**Comment by Viicos** [comment]
This is now implemented with https://github.com/pydantic/pydantic/pull/6563, which will probably land in 2.7.0.
